### PR TITLE
Fuse_modules in a qat-respecting way

### LINKED ISSUE #12890: `QuantizationAwareTraining` does not use `torch.ao.quantization.fuse_modules_qat`
## 🐛 Bug
Models in `QuantizationAwareTraining` should be fused in a qat-respecting manner, but currently are not

### To Reproduce
Look at the code, where QAT uses [`torch.quantization.fuse_modules`](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/callbacks/quantization.py#L255), which calls [`torch.ao.quantization._fuse_modules`]((https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/callbacks/quantization.py#L255)) with `qat=False` internally.

### Expected behavior
Instead, QAT should call [`torch.ao.quantization.fuse_modules_qat`](https://github.com/pytorch/pytorch/blob/bc2c6edaf163b1a1330e37a6e34caf8c553e4755/torch/ao/quantization/fuse_modules.py#L158-L167), which correctly calls `torch.ao.quantization._fuse_modules` with `is_qat=True`

### Additional context
Should torch map `torch.ao.quantization.fuse_modules_qat` to `torch.quantization.fuse_modules_qat` in future, code should be refactored similar to [QConfig](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/callbacks/quantization.py#L32-L35).

cc @borda @Borda

### PR DESCRIPTION
**Pull request by ORippler**
## What does this PR do?
For QAT, we should fuse modules in a qat-respecting way

Fixes #12890

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/quantization.py [hunk 1]
```diff
@@ -252,7 +252,7 @@ def _prepare_model(self, model: torch.nn.Module) -> None:
             model.qconfig = self._qconfig
 
         if self._check_feasible_fuse(model):
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:254]: Can you update the CHANGELOG?

File: pytorch_lightning/callbacks/quantization.py [hunk 2]
```diff
@@ -252,7 +252,7 @@ def _prepare_model(self, model: torch.nn.Module) -> None:
...
-            torch.quantization.fuse_modules(model, self._modules_to_fuse, inplace=True)
+            torch.ao.quantization.fuse_modules_qat(model, self._modules_to_fuse, inplace=True)
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:255]: The linked issue says: 

> Once torch maps torch.ao.quantization.fuse_modules_qat to torch.quantization.fuse_modules_qat, code should be refactored

Do you know when this will happen?

**Review by carmocca — APPROVED** [review]
Do you think we could test this change?

File: pytorch_lightning/callbacks/quantization.py (hunk 2 above)
Comment (akihironitta) [inline pytorch_lightning/callbacks/quantization.py:255]: AFAIK, `torch.ao` is available in PyTorch 1.10 or later, so it needs to be protected as we protect QConfig:
https://github.com/PyTorchLightning/pytorch-lightning/blob/cac02a04ba78a4bb69648f2918dbbf5798d5d84f/pytorch_lightning/callbacks/quantization.py#L32-L35

Is there a function that is equivalent to `torch.ao.quantization.fuse_modules_qat` in PyTorch 1.8 or 1.9?

Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:255]: @ORippler ^^ :rabbit: 
we would need t make it back compatible to 1.8

Comment (ORippler) [reply pytorch_lightning/callbacks/quantization.py]: I looked up in the release notes, and for `pytorch<=1.10`, `torch.quantization.fuse_modules` implicitly uses the `Module.training` flag to determine whether fusion should happen in a QAT aware manner. I restored backwards compatibility by dynamic imports

File: pytorch_lightning/callbacks/quantization.py (hunk 1 above)
Comment (ORippler) [reply pytorch_lightning/callbacks/quantization.py:255]: Sure :)

File: pytorch_lightning/callbacks/quantization.py (hunk 2 above)
Comment (ORippler) [reply pytorch_lightning/callbacks/quantization.py]: > The linked issue says:
> 
> > Once torch maps torch.ao.quantization.fuse_modules_qat to torch.quantization.fuse_modules_qat, code should be refactored
> 
> Do you know when this will happen?

No clue, sorry, even the part that it will be remapped by torch is a guess on my part, refactored the issue

File: CHANGELOG.md [hunk 3]
```diff
@@ -141,6 +141,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
+- Fixed `fuse_modules` to be qat-aware for `pytorch>1.10`. ([#12891](https://github.com/PyTorchLightning/pytorch-lightning/pull/12891))
```
Comment (akihironitta) [inline CHANGELOG.md:144]: The change here impacts 1.10 as well, right?
```suggestion
- Fixed `fuse_modules` to be qat-aware for `torch>=1.10`. ([#12891](https://github.com/PyTorchLightning/pytorch-lightning/pull/12891))
```

File: pytorch_lightning/callbacks/quantization.py (hunk 2 above)
Comment (akihironitta) [reply pytorch_lightning/callbacks/quantization.py]: Thank you for your action!

File: CHANGELOG.md (hunk 3 above)
Comment (akihironitta) [reply CHANGELOG.md:144]: Just realised from your comment and one of your commits https://github.com/PyTorchLightning/pytorch-lightning/pull/12891/commits/c122790c7540e0721f7eeabcde3d2d0ffc9c06e1 that ` torch.ao.quantization.fuse_modules_qat` is only available in PyTorch 1.11+. Sorry for the confusion!

Comment (akihironitta) [inline CHANGELOG.md:144]:
```suggestion
- Fixed `fuse_modules` to be qat-aware for `torch>=1.11`. ([#12891](https://github.com/PyTorchLightning/pytorch-lightning/pull/12891))
```

**Review by akihironitta — APPROVED** [review]

**Review by otaj — APPROVED** [review]
